src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+    *data_length = ntoh32u(drive->rd_buf);
+    drive->cmd[7] = ((*data_length + 4) >> 8) & 0xFF;
+    drive->cmd[8] = (*data_length + 4) & 0xFF;
+    drive->cmd[9] = 0;
+    if (*data_length > 4)
+      if ((drive->err =
+               drive->cmd.transport(READ, drive->rd_buf, *data_length + 4))) {
+        if (!drive->silent)
+          sperror("GET_CONFIGURATION", drive->err);
+        return (drive->err);
+      }
+  }
+  if (current)
+    *current = drive->rd_buf[10] & 0x01;
+  return 0;
+}
+
+void detect_iface(drive_info *drive) {
+  unsigned int len = 0;
+  get_configuration(drive, FEATURE_CORE, &len, NULL);
+  drive->iface_id = (drive->rd_buf[12] << 12) | (drive->rd_buf[13] << 8) |
+                    (drive->rd_buf[14] << 4) | drive->rd_buf[15];
+  if (drive->iface_id < iface_id_max)
+    strcpy(drive->iface, iface_list[drive->iface_id]);
+  else
+    strcpy(drive->iface, iface_list[iface_id_max + 1]);
+}
+
+int write_buffer(drive_info *drive, uint8_t mode, uint8_t buff_id,
+                 uint32_t offs, uint32_t len) {
+  drive->cmd[0] = SPC_WRITE_BUFFER;
+  drive->cmd[1] = (mode & 0x0F);
+  drive->cmd[2] = buff_id;
+
+  drive->cmd[3] = (offs >> 16) & 0xFF;
+  drive->cmd[4] = (offs >> 8) & 0xFF;
+  drive->cmd[5] = (offs)&0xFF;
+
+  drive->cmd[6] = (len >> 16) & 0xFF;
+  drive->cmd[7] = (len >> 8) & 0xFF;
+  drive->cmd[8] = (len)&0xFF;
+  drive->cmd[9] = 0;
+
+  if ((drive->err = drive->cmd.transport(WRITE, drive->rd_buf, len))) {
+    if (!drive->silent)
+      sperror("WRITE_BUFFER", drive->err);
+    return (drive->err);
+  }
+
+  return 0;
+}
+
+int read_buffer(drive_info *drive, uint8_t mode, uint8_t buff_id, uint32_t offs,
+                uint32_t len) {
+  printf("read buffer: mode %x, id %x, offs %x, len %x\n", mode, buff_id, offs,
+         len);
+  drive->cmd[0] = SPC_READ_BUFFER;
+  drive->cmd[1] = (mode & 0x0F);
+  drive->cmd[2] = buff_id;
+
+  drive->cmd[3] = (offs >> 16) & 0xFF;
+  drive->cmd[4] = (offs >> 8) & 0xFF;
+  drive->cmd[5] = (offs)&0xFF;
+
+  drive->cmd[6] = (len >> 16) & 0xFF;
+  drive->cmd[7] = (len >> 8) & 0xFF;
+  drive->cmd[8] = (len)&0xFF;
+  drive->cmd[9] = 0;
+
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, len))) {
+    if (!drive->silent)
+      sperror("READ_BUFFER", drive->err);
+    return (drive->err);
+  }
+
+  return 0;
+}
+
+int read_echo_buffer_size(drive_info *drive) {
+  drive->cmd[0] = SPC_READ_BUFFER;
+  //	drive->cmd[1] = 0x00;
+  drive->cmd[1] = 0x0B; // echo buffer descriptor
+  drive->cmd[8] = 4;
+  drive->cmd[9] = 0;
+
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, 4))) {
+    if (!drive->silent)
+      sperror("READ_ECHO_BUFFER_SIZE", drive->err);
+    return (0);
+  }
+
+  return ntoh16(drive->rd_buf + 2) & 0x1FFF;
+}
+
+int test_dma_speed(drive_info *drive, long msecs) {
+  long st, et;
+  long speed = 0;
+
+  int cnt = 0;
+  drive->silent--;
+
+  uint32_t buf_size = read_echo_buffer_size(drive);
+  printf("Echo buffer size: %d\n", buf_size);
```

Comment:
Codacy found an issue: [%d in format string (no. 1) requires 'int' but the argument type is 'unsigned int'.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)